snuba/subscriptions/worker.py

```diff
@@ -2,47 +2,85 @@
 from snuba.subscriptions.data import Subscription
-from snuba.subscriptions.executor import SubscriptionExecutor
 from snuba.subscriptions.scheduler import ScheduledTask, Scheduler
 from snuba.utils.codecs import Codec
+from snuba.utils.metrics.backends.abstract import MetricsBackend
+from snuba.utils.metrics.gauge import Gauge
+from snuba.utils.metrics.timer import Timer
 from snuba.utils.streams.batching import AbstractBatchWorker
 from snuba.utils.streams.kafka import KafkaPayload
 from snuba.utils.streams.producer import Producer
 from snuba.utils.streams.types import Message, Topic
+from snuba.web.query import parse_and_run_query
 
 
-class SubscriptionResultFuture(NamedTuple):
+class SubscriptionTaskResultFuture(NamedTuple):
     task: ScheduledTask[Subscription]
-    future: Future[Result]
+    future: Future[Tuple[Request, Result]]
 
 
-class SubscriptionResult(NamedTuple):
+class SubscriptionTaskResult(NamedTuple):
     task: ScheduledTask[Subscription]
-    result: Result
+    result: Tuple[Request, Result]
 
 
-class SubscriptionWorker(AbstractBatchWorker[Tick, Sequence[SubscriptionResultFuture]]):
+class SubscriptionWorker(
+    AbstractBatchWorker[Tick, Sequence[SubscriptionTaskResultFuture]]
+):
     def __init__(
         self,
-        executor: SubscriptionExecutor,
+        dataset: Dataset,
+        executor: ThreadPoolExecutor,
         schedulers: Mapping[int, Scheduler[Subscription]],
-        producer: Producer[SubscriptionResult],
+        producer: Producer[SubscriptionTaskResult],
         topic: Topic,
+        metrics: MetricsBackend,
     ) -> None:
+        self.__dataset = dataset
         self.__executor = executor
         self.__schedulers = schedulers
         self.__producer = producer
         self.__topic = topic
+        self.__metrics = metrics
+
+        self.__concurrent_gauge = Gauge(self.__metrics, "executor.concurrent")
+
+    def __execute(
+        self, task: ScheduledTask[Subscription], tick: Tick
+    ) -> Tuple[Request, Result]:
+        # Measure the amount of time that took between this task being
+        # scheduled and it beginning to execute.
+        self.__metrics.timing(
+            "executor.latency", (time.time() - task.timestamp.timestamp()) * 1000
+        )
+
+        # XXX: The ``query`` name is taken from the web views so that all query
+        # performance metrics are reported to the same spot, regardless of
+        # execution environment.
+        timer = Timer("query")
+
+        request = task.task.data.build_request(
+            self.__dataset, task.timestamp, tick.offsets.upper, timer,
+        )
+
+        with self.__concurrent_gauge:
+            # XXX: The ``extra`` is discarded from ``QueryResult`` since it is
+            # not particularly useful in this context and duplicates data that
+            # is already being published to the query log.
+            return request, parse_and_run_query(self.__dataset, request, timer).result
```

Comment:
I am not sure what the request object in the response is meant to be used for so I am not sure this comment is pertinent. Pay attention, this is not the request you would send to Snuba to run the query. The object you pass to parse_and_run_query is mutable and it gets mutated during the query execution. The object you get from request.body is taken from request.query.body, which is mutated during query execution.
Specifically:
- query extensions are tweaking the query object (the query body in the legacy representation) in place
- dataset query processors are still mutating the query in place in the request.
You do not care about clickhouse query processor since, after #880, we do a copy during translation and the query from that point on is isolated.

In general, as long as the request object contains mutable objects I think it would be safer to return a copy of the request. Of course unless the processed query is what you want to return.

Reply:
> I am not sure what the request object in the response is meant to be used for so I am not sure this comment is pertinent.

It's described in the description, but more detail: this is to allow us to double check the query being sent to the database via the subscription consumer is what we expected, and also to allow the same request to be replayed later to ensure that the value hasn't changed. This also would allow the consumer to selectively extract the conditions and to identify the specific data points under aggregation in the subscription (no specific use case for this yet, but we've discussed this being useful in the past.)

I don't think whether this happens before or after query processing doesn't really matter that much to the end consumer — but since there are no guarantees that the processors themselves are idempotent, I agree that it's probably safer for that reason to copy the original request here after it's created to ensure they're not re-applied in a way that might materially change the query.

Ultimately, I don't think this is a great solution for "this is to allow us to double check the query being sent to the database via the subscription consumer is what we expected" but it's the approach with the least entanglement to anything else right now.